waddrmgr/scoped_manager.go

```diff
@@ -2029,16 +2053,63 @@ func (s *ScopedKeyManager) toImportedPublicManagedAddress(
 func (s *ScopedKeyManager) ImportScript(ns walletdb.ReadWriteBucket,
 	script []byte, bs *BlockStamp) (ManagedScriptAddress, error) {
 
+	return s.importScriptAddress(ns, script, bs, Script, 0, true)
+}
+
+// ImportWitnessScript imports a user-provided script into the address manager.
+// The imported script will act as a pay-to-witness-script-hash address.
+//
+// All imported script addresses will be part of the account defined by the
+// ImportedAddrAccount constant.
+//
+// When the address manager is watching-only, the script itself will not be
+// stored or available since it is considered private data.
+//
+// This function will return an error if the address manager is locked and not
+// watching-only, or the address already exists.  Any other errors returned are
+// generally unexpected.
+func (s *ScopedKeyManager) ImportWitnessScript(ns walletdb.ReadWriteBucket,
+	script []byte, bs *BlockStamp, witnessVersion byte,
+	isSecretScript bool) (ManagedScriptAddress, error) {
+
+	return s.importScriptAddress(
+		ns, script, bs, WitnessScript, witnessVersion, isSecretScript,
+	)
+}
+
+// importScriptAddress imports a new pay-to-script or pay-to-witness-script
+// address.
+func (s *ScopedKeyManager) importScriptAddress(ns walletdb.ReadWriteBucket,
+	script []byte, bs *BlockStamp, addrType AddressType,
+	witnessVersion byte, isSecretScript bool) (ManagedScriptAddress,
+	error) {
+
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
 
 	// The manager must be unlocked to encrypt the imported script.
-	if s.rootManager.IsLocked() {
+	if isSecretScript && s.rootManager.IsLocked() {
 		return nil, managerError(ErrLocked, errLocked, nil)
 	}
 
+	// A secret script can only be used with a non-watch only manager. If
+	// a wallet is watch-only then the script must be encrypted with the
+	// public encryption key.
+	if isSecretScript && s.rootManager.WatchOnly() {
+		return nil, managerError(ErrWatchingOnly, errWatchingOnly, nil)
+	}
+
+	// Witness script addresses use a SHA256.
+	var scriptHash []byte
+	switch addrType {
+	case WitnessScript:
```

Comment:
We'd need to modify this if we _wanted_ to support nested P2SH on this level, but I don't think we have a direct need for that, and can simplify the implementation by leaving that out.